File: viddhana/packages/sdks/javascript/src/modules/atlas.ts
/**
 * Atlas Chain Module - Chain interaction methods
 */

import { ethers } from 'ethers';
import type { ChainInfo, Block, TransactionReceipt } from '../types';

/**
 * Atlas chain module for interacting with the VIDDHANA blockchain
 *
 * @example
 * ```typescript
 * const chainInfo = await client.atlas.getChainInfo();
 * console.log(`Block: ${chainInfo.blockNumber}`);
 *
 * const balance = await client.atlas.getBalance('0x...');
 * console.log(`Balance: ${ethers.formatEther(balance)} VDH`);
 * ```
 */
export class AtlasChain {
  private readonly provider: ethers.JsonRpcProvider;
  private readonly signer: ethers.Wallet | null;

  constructor(
    provider: ethers.JsonRpcProvider,
    signer: ethers.Wallet | null
  ) {
    this.provider = provider;
    this.signer = signer;
  }

  /**
   * Gets chain information
   *
   * @returns Chain info including chainId, name, block number, and gas price
   */
  async getChainInfo(): Promise<ChainInfo> {
    const [network, blockNumber, feeData] = await Promise.all([
      this.provider.getNetwork(),
      this.provider.getBlockNumber(),
      this.provider.getFeeData(),
    ]);

    return {
      chainId: Number(network.chainId),
      name: 'VIDDHANA Atlas',
      symbol: 'VDH',
      blockNumber,
      gasPrice: feeData.gasPrice?.toString() ?? '0',
    };
  }

  /**
   * Gets the balance of an address
   *
   * @param address - The address to check
   * @returns Balance in wei as bigint
   */
  async getBalance(address: string): Promise<bigint> {
    return this.provider.getBalance(address);
  }

  /**
   * Gets the formatted balance in VDH
   *
   * @param address - The address to check
   * @returns Balance formatted as string with decimals
   */
  async getFormattedBalance(address: string): Promise<string> {
    const balance = await this.getBalance(address);
    return ethers.formatEther(balance);
  }

  /**
   * Gets a block by number or hash
   *
   * @param blockHashOrNumber - Block number or hash
   * @returns Block information
   */
  async getBlock(
    blockHashOrNumber: number | string = 'latest'
  ): Promise<Block | null> {
    const block = await this.provider.getBlock(blockHashOrNumber);

    if (!block) {
      return null;
    }

    return {
      number: block.number,
      hash: block.hash ?? '',
      parentHash: block.parentHash,
      timestamp: block.timestamp,
      nonce: block.nonce,
      difficulty: BigInt(0), // PoS chains don't have difficulty
      gasLimit: block.gasLimit,
      gasUsed: block.gasUsed,
      miner: block.miner,
      transactions: block.transactions.map((tx) =>
        typeof tx === 'string' ? tx : tx.hash
      ),
    };
  }

  /**
   * Gets the latest block number
   *
   * @returns Current block number
   */
  async getBlockNumber(): Promise<number> {
    return this.provider.getBlockNumber();
  }

  /**
   * Gets a transaction receipt
   *
   * @param txHash - Transaction hash
   * @returns Transaction receipt or null if not found
   */
  async getTransactionReceipt(
    txHash: string
  ): Promise<TransactionReceipt | null> {
    const receipt = await this.provider.getTransactionReceipt(txHash);

    if (!receipt) {
      return null;
    }

    return {
      transactionHash: receipt.hash,
      blockNumber: receipt.blockNumber,
      blockHash: receipt.blockHash,
      from: receipt.from,
      to: receipt.to ?? '',
      gasUsed: receipt.gasUsed,
      status: receipt.status ?? 0,
      logs: receipt.logs.map((log) => ({
        address: log.address,
        topics: log.topics as string[],
        data: log.data,
        blockNumber: log.blockNumber,
        transactionHash: log.transactionHash,
        logIndex: log.index,
      })),
    };
  }

  /**
   * Gets the current gas price
   *
   * @returns Gas price in wei
   */
  async getGasPrice(): Promise<bigint> {
    const feeData = await this.provider.getFeeData();
    return feeData.gasPrice ?? 0n;
  }

  /**
   * Estimates gas for a transaction
   *
   * @param tx - Transaction request
   * @returns Estimated gas limit
   */
  async estimateGas(tx: ethers.TransactionRequest): Promise<bigint> {
    return this.provider.estimateGas(tx);
  }

  /**
   * Gets the transaction count (nonce) for an address
   *
   * @param address - The address to check
   * @returns Transaction count
   */
  async getTransactionCount(address: string): Promise<number> {
    return this.provider.getTransactionCount(address);
  }

  /**
   * Sends a signed transaction
   *
   * @param signedTx - Signed transaction data
   * @returns Transaction response
   */
  async sendTransaction(
    signedTx: string
  ): Promise<ethers.TransactionResponse> {
    return this.provider.broadcastTransaction(signedTx);
  }

  /**
   * Waits for a transaction to be confirmed
   *
   * @param txHash - Transaction hash
   * @param confirmations - Number of confirmations to wait for
   * @returns Transaction receipt
   */
  async waitForTransaction(
    txHash: string,
    confirmations = 1
  ): Promise<TransactionReceipt | null> {
    const receipt = await this.provider.waitForTransaction(
      txHash,
      confirmations
    );

    if (!receipt) {
      return null;
    }

    return {
      transactionHash: receipt.hash,
      blockNumber: receipt.blockNumber,
      blockHash: receipt.blockHash,
      from: receipt.from,
      to: receipt.to ?? '',
      gasUsed: receipt.gasUsed,
      status: receipt.status ?? 0,
      logs: receipt.logs.map((log) => ({
        address: log.address,
        topics: log.topics as string[],
        data: log.data,
        blockNumber: log.blockNumber,
        transactionHash: log.transactionHash,
        logIndex: log.index,
      })),
    };
  }

  /**
   * Gets tokenomics statistics from the chain
   *
   * @returns Tokenomics statistics
   */
  async getTokenomicsStats(): Promise<Record<string, unknown>> {
    const result = await this.provider.send('vdh_getTokenomicsStats', []);
    return result as Record<string, unknown>;
  }

  /**
   * Subscribes to new blocks
   *
   * @param callback - Function to call on each new block
   * @returns Unsubscribe function
   */
  onBlock(callback: (blockNumber: number) => void): () => void {
    this.provider.on('block', callback);
    return () => this.provider.off('block', callback);
  }
}

File: viddhana/packages/sdks/python/src/viddhana/atlas.py
"""Atlas Chain module for blockchain interactions."""

from typing import Optional, TYPE_CHECKING
from web3 import Web3
from web3.types import Wei

from .types import ChainInfo, BlockInfo

if TYPE_CHECKING:
    from eth_account.signers.local import LocalAccount


class AtlasModule:
    """Module for interacting with the Atlas blockchain."""

    def __init__(self, w3: Web3, account: Optional["LocalAccount"] = None):
        """
        Initialize AtlasModule.

        Args:
            w3: Web3 instance connected to Atlas chain.
            account: Optional account for signing transactions.
        """
        self._w3 = w3
        self._account = account

    def get_chain_info(self) -> ChainInfo:
        """
        Get current chain information.

        Returns:
            ChainInfo with chain ID, block number, gas price, and network name.
        """
        chain_id = self._w3.eth.chain_id
        block_number = self._w3.eth.block_number
        gas_price = self._w3.eth.gas_price

        # Determine network name based on chain ID
        network_names = {
            13370: "viddhana-mainnet",
            13371: "viddhana-testnet",
        }
        network_name = network_names.get(chain_id, f"unknown-{chain_id}")

        return ChainInfo(
            chain_id=chain_id,
            block_number=block_number,
            gas_price=gas_price,
            network_name=network_name,
        )

    def get_balance(self, address: str) -> Wei:
        """
        Get the native token balance of an address.

        Args:
            address: The address to query.

        Returns:
            Balance in Wei.
        """
        checksum_address = Web3.to_checksum_address(address)
        return self._w3.eth.get_balance(checksum_address)

    def get_balance_ether(self, address: str) -> float:
        """
        Get the native token balance in Ether units.

        Args:
            address: The address to query.

        Returns:
            Balance in Ether (float).
        """
        balance_wei = self.get_balance(address)
        return float(Web3.from_wei(balance_wei, "ether"))

    def get_block(self, block_identifier: int | str = "latest") -> BlockInfo:
        """
        Get block information.

        Args:
            block_identifier: Block number or 'latest', 'pending', 'earliest'.

        Returns:
            BlockInfo with block details.
        """
        block = self._w3.eth.get_block(block_identifier)

        return BlockInfo(
            number=block["number"],
            hash=block["hash"].hex(),
            parent_hash=block["parentHash"].hex(),
            timestamp=block["timestamp"],
            transactions=[tx.hex() if isinstance(tx, bytes) else tx for tx in block["transactions"]],
            gas_used=block["gasUsed"],
            gas_limit=block["gasLimit"],
        )

    def get_transaction_count(self, address: str) -> int:
        """
        Get the transaction count (nonce) for an address.

        Args:
            address: The address to query.

        Returns:
            Transaction count.
        """
        checksum_address = Web3.to_checksum_address(address)
        return self._w3.eth.get_transaction_count(checksum_address)

    def get_gas_price(self) -> int:
        """
        Get current gas price.

        Returns:
            Gas price in Wei.
        """
        return self._w3.eth.gas_price

    def is_connected(self) -> bool:
        """
        Check if connected to the network.

        Returns:
            True if connected, False otherwise.
        """
        return self._w3.is_connected()
